[Flaky test] MultiStageEngineCustomTenantIntegrationTest.testGeneratedQueries()

```
Error:  org.apache.pinot.integration.tests.MultiStageEngineCustomTenantIntegrationTest.testGeneratedQueries -- Time elapsed: 14.29 s <<< FAILURE!
java.lang.AssertionError: 
Caught exception while testing query!
Pinot query: SELECT "ArrTime", "WheelsOff", "Cancelled" FROM mytable WHERE "OriginAirportID" > 12177 OR "TaxiIn" BETWEEN 47 AND 70 AND "FlightNum" = 1077 LIMIT 10000
H2 query: SELECT `ArrTime`, `WheelsOff`, `Cancelled` FROM mytable WHERE `OriginAirportID` > 12177 OR `TaxiIn` BETWEEN 47 AND 70 AND `FlightNum` = 1077 LIMIT 10000
	at org.testng.Assert.fail(Assert.java:99)
	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.failure(ClusterIntegrationTestUtils.java:1076)
	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.failure(ClusterIntegrationTestUtils.java:1060)
	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.testQuery(ClusterIntegrationTestUtils.java:699)
	at org.apache.pinot.integration.tests.BaseClusterIntegrationTest.testQuery(BaseClusterIntegrationTest.java:739)
	at org.apache.pinot.integration.tests.BaseClusterIntegrationTestSet.testGeneratedQueries(BaseClusterIntegrationTestSet.java:485)
	at org.apache.pinot.integration.tests.MultiStageEngineIntegrationTest.testGeneratedQueries(MultiStageEngineIntegrationTest.java:132)
Caused by: org.apache.pinot.client.PinotClientException: Query had processing exceptions: 
[{"errorCode":200,"message":"QueryExecutionError:\nReceived error query execution result block: {450=InternalError:\njava.util.ConcurrentModificationException\n\tat java.base/java.util.ArrayList$Itr.checkForComodification(ArrayList.java:1043)\n\tat java.base/java.util.ArrayList$Itr.next(ArrayList.java:997)\n\tat org.apache.pinot.core.operator.docidsets.AndDocIdSet.getNumEntriesScannedInFilter(AndDocIdSet.java:189)\n\tat org.apache.pinot.core.operator.docidsets.OrDocIdSet.getNumEntriesScannedInFilter(OrDocIdSet.java:131)}\norg.apache.pinot.query.service.dispatch.QueryDispatcher.getResultTable(QueryDispatcher.java:315)\norg.apache.pinot.query.service.dispatch.QueryDispatcher.runReducer(QueryDispatcher.java:255)\norg.apache.pinot.query.service.dispatch.QueryDispatcher.submitAndReduce(QueryDispatcher.java:99)\norg.apache.pinot.broker.requesthandler.MultiStageBrokerRequestHandler.handleRequest(MultiStageBrokerRequestHandler.java:191)\n"}]
	at org.apache.pinot.client.Connection.execute(Connection.java:103)
	at org.apache.pinot.client.Connection.execute(Connection.java:83)
	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.testQueryInternal(ClusterIntegrationTestUtils.java:736)
	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.testQuery(ClusterIntegrationTestUtils.java:696)
	... 36 more
```

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/docidsets/AndDocIdSet.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/docidsets/AndDocIdSet.java
@@ -20,11 +20,10 @@ package org.apache.pinot.core.operator.docidsets;
 
 import java.util.ArrayList;
 import java.util.Comparator;
-import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
 import javax.annotation.Nullable;
-import org.apache.commons.collections.MapUtils;
 import org.apache.pinot.common.utils.config.QueryOptionsUtils;
 import org.apache.pinot.core.common.BlockDocIdIterator;
 import org.apache.pinot.core.common.BlockDocIdSet;
@@ -57,14 +56,16 @@ import org.roaringbitmap.buffer.MutableRoaringBitmap;
  * </ul>
  */
 public final class AndDocIdSet implements BlockDocIdSet {
-  private final List<BlockDocIdSet> _docIdSets;
+  // Keep the scan based BlockDocIdSets to be accessed when collecting query execution stats
+  private final AtomicReference<List<BlockDocIdSet>> _scanBasedDocIdSets = new AtomicReference<>();
   private final boolean _cardinalityBasedRankingForScan;
-  private long _numEntriesScannedInFilter = 0L;
+  private List<BlockDocIdSet> _docIdSets;
+  private volatile long _numEntriesScannedInFilter;
 
   public AndDocIdSet(List<BlockDocIdSet> docIdSets, @Nullable Map<String, String> queryOptions) {
-    _docIdSets = docIdSets instanceof ArrayList ? docIdSets : new ArrayList<>(docIdSets);
+    _docIdSets = docIdSets;
     _cardinalityBasedRankingForScan =
-        !MapUtils.isEmpty(queryOptions) && QueryOptionsUtils.isAndScanReorderingEnabled(queryOptions);
+        queryOptions != null && QueryOptionsUtils.isAndScanReorderingEnabled(queryOptions);
   }
 
   @Override
@@ -77,32 +78,32 @@ public final class AndDocIdSet implements BlockDocIdSet {
     List<BitmapBasedDocIdIterator> bitmapBasedDocIdIterators = new ArrayList<>();
     List<ScanBasedDocIdIterator> scanBasedDocIdIterators = new ArrayList<>();
     List<BlockDocIdIterator> remainingDocIdIterators = new ArrayList<>();
+    long numEntriesScannedForNonScanBasedDocIdSets = 0L;
+    List<BlockDocIdSet> scanBasedDocIdSets = new ArrayList<>();
 
-    Iterator<BlockDocIdSet> iterator = _docIdSets.iterator();
-    for (int i = 0; iterator.hasNext(); i++) {
-      BlockDocIdSet blockDocIdSet = iterator.next();
-      BlockDocIdIterator docIdIterator = blockDocIdSet.iterator();
+    for (int i = 0; i < numDocIdSets; i++) {
+      BlockDocIdSet docIdSet = _docIdSets.get(i);
+      BlockDocIdIterator docIdIterator = docIdSet.iterator();
       allDocIdIterators[i] = docIdIterator;
       if (docIdIterator instanceof SortedDocIdIterator) {
         sortedDocIdIterators.add((SortedDocIdIterator) docIdIterator);
-        // aggregate the number of entries scanned in filter before removing the iterator
-        _numEntriesScannedInFilter += blockDocIdSet.getNumEntriesScannedInFilter();
-        // do not keep holding on to the _docIdRanges since they will occupy heap space during the query execution
-        iterator.remove();
+        numEntriesScannedForNonScanBasedDocIdSets += docIdSet.getNumEntriesScannedInFilter();
       } else if (docIdIterator instanceof BitmapBasedDocIdIterator) {
         bitmapBasedDocIdIterators.add((BitmapBasedDocIdIterator) docIdIterator);
-        // aggregate the number of entries scanned in filter before removing the iterator
-        // some BitmapBasedDocIdIterator may be generated from underlying index types (e.g. H3Index) that actually
-        // scans documents, so we need to aggregate them here
-        _numEntriesScannedInFilter += blockDocIdSet.getNumEntriesScannedInFilter();
-        // do not keep holding on to the _docIdRanges since they will occupy heap space during the query execution
-        iterator.remove();
+        numEntriesScannedForNonScanBasedDocIdSets += docIdSet.getNumEntriesScannedInFilter();
       } else if (docIdIterator instanceof ScanBasedDocIdIterator) {
         scanBasedDocIdIterators.add((ScanBasedDocIdIterator) docIdIterator);
+        scanBasedDocIdSets.add(docIdSet);
       } else {
         remainingDocIdIterators.add(docIdIterator);
+        scanBasedDocIdSets.add(docIdSet);
       }
     }
+
+    // Set _docIdSets to null so that underlying BlockDocIdSets can be garbage collected
+    _docIdSets = null;
+    _numEntriesScannedInFilter = numEntriesScannedForNonScanBasedDocIdSets;
+    _scanBasedDocIdSets.set(scanBasedDocIdSets);
 
     // evaluate the bitmaps in the order of the lowest matching num docIds comes first, so that we minimize the number
     // of containers (range) for comparison from the beginning, as will minimize the effort of bitmap AND application
@@ -186,9 +187,13 @@ public final class AndDocIdSet implements BlockDocIdSet {
 
   @Override
   public long getNumEntriesScannedInFilter() {
-    for (BlockDocIdSet child : _docIdSets) {
-      _numEntriesScannedInFilter += child.getNumEntriesScannedInFilter();
+    List<BlockDocIdSet> scanBasedDocIdSets = _scanBasedDocIdSets.get();
+    long numEntriesScannedForScanBasedDocIdSets = 0L;
+    if (scanBasedDocIdSets != null) {
+      for (BlockDocIdSet scanBasedDocIdSet : scanBasedDocIdSets) {
+        numEntriesScannedForScanBasedDocIdSets += scanBasedDocIdSet.getNumEntriesScannedInFilter();
+      }
     }
-    return _numEntriesScannedInFilter;
+    return _numEntriesScannedInFilter + numEntriesScannedForScanBasedDocIdSets;
   }
 }

--- a/pinot-core/src/main/java/org/apache/pinot/core/operator/docidsets/OrDocIdSet.java
+++ b/pinot-core/src/main/java/org/apache/pinot/core/operator/docidsets/OrDocIdSet.java
@@ -19,8 +19,8 @@
 package org.apache.pinot.core.operator.docidsets;
 
 import java.util.ArrayList;
-import java.util.Iterator;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
 import org.apache.pinot.core.common.BlockDocIdIterator;
 import org.apache.pinot.core.common.BlockDocIdSet;
 import org.apache.pinot.core.operator.dociditerators.BitmapBasedDocIdIterator;
@@ -48,12 +48,14 @@ import org.roaringbitmap.buffer.MutableRoaringBitmap;
  * </ul>
  */
 public final class OrDocIdSet implements BlockDocIdSet {
-  private final List<BlockDocIdSet> _docIdSets;
+  // Keep the scan based BlockDocIdSets to be accessed when collecting query execution stats
+  private final AtomicReference<List<BlockDocIdSet>> _scanBasedDocIdSets = new AtomicReference<>();
   private final int _numDocs;
-  private long _numEntriesScannedInFilter = 0L;
+  private List<BlockDocIdSet> _docIdSets;
+  private volatile long _numEntriesScannedInFilter = 0L;
 
   public OrDocIdSet(List<BlockDocIdSet> docIdSets, int numDocs) {
-    _docIdSets = docIdSets instanceof ArrayList ? docIdSets : new ArrayList<>(docIdSets);
+    _docIdSets = docIdSets;
     _numDocs = numDocs;
   }
 
@@ -64,30 +66,29 @@ public final class OrDocIdSet implements BlockDocIdSet {
     List<SortedDocIdIterator> sortedDocIdIterators = new ArrayList<>();
     List<BitmapBasedDocIdIterator> bitmapBasedDocIdIterators = new ArrayList<>();
     List<BlockDocIdIterator> remainingDocIdIterators = new ArrayList<>();
+    long numEntriesScannedForNonScanBasedDocIdSets = 0L;
+    List<BlockDocIdSet> scanBasedDocIdSets = new ArrayList<>();
 
-    Iterator<BlockDocIdSet> iterator = _docIdSets.iterator();
-    for (int i = 0; iterator.hasNext(); i++) {
-      BlockDocIdSet blockDocIdSet = iterator.next();
-      BlockDocIdIterator docIdIterator = blockDocIdSet.iterator();
+    for (int i = 0; i < numDocIdSets; i++) {
+      BlockDocIdSet docIdSet = _docIdSets.get(i);
+      BlockDocIdIterator docIdIterator = docIdSet.iterator();
       allDocIdIterators[i] = docIdIterator;
       if (docIdIterator instanceof SortedDocIdIterator) {
         sortedDocIdIterators.add((SortedDocIdIterator) docIdIterator);
-        // aggregate the number of entries scanned in filter before removing the iterator
-        _numEntriesScannedInFilter += blockDocIdSet.getNumEntriesScannedInFilter();
-        // do not keep holding on to the _docIdRanges since they will occupy heap space during the query execution
-        iterator.remove();
+        numEntriesScannedForNonScanBasedDocIdSets += docIdSet.getNumEntriesScannedInFilter();
       } else if (docIdIterator instanceof BitmapBasedDocIdIterator) {
-        bitmapBasedDocIdIterators.add((BitmapBasedDocIdIterator) docIdIterator);
-        // aggregate the number of entries scanned in filter before removing the iterator
-        // some BitmapBasedDocIdIterator may be generated from underlying index types (e.g. H3Index) that actually
-        // scans documents, so we need to aggregate them here
-        _numEntriesScannedInFilter += blockDocIdSet.getNumEntriesScannedInFilter();
-        // do not keep holding on to the bitmaps since they will occupy heap space during the query execution
-        iterator.remove();
+        numEntriesScannedForNonScanBasedDocIdSets += docIdSet.getNumEntriesScannedInFilter();
       } else {
         remainingDocIdIterators.add(docIdIterator);
+        scanBasedDocIdSets.add(docIdSet);
       }
     }
+
+    // Set _docIdSets to null so that underlying BlockDocIdSets can be garbage collected
+    _docIdSets = null;
+    _numEntriesScannedInFilter = numEntriesScannedForNonScanBasedDocIdSets;
+    _scanBasedDocIdSets.set(scanBasedDocIdSets);
+
     int numSortedDocIdIterators = sortedDocIdIterators.size();
     int numBitmapBasedDocIdIterators = bitmapBasedDocIdIterators.size();
     if (numSortedDocIdIterators + numBitmapBasedDocIdIterators > 1) {
@@ -127,9 +128,13 @@ public final class OrDocIdSet implements BlockDocIdSet {
 
   @Override
   public long getNumEntriesScannedInFilter() {
-    for (BlockDocIdSet docIdSet : _docIdSets) {
-      _numEntriesScannedInFilter += docIdSet.getNumEntriesScannedInFilter();
+    List<BlockDocIdSet> scanBasedDocIdSets = _scanBasedDocIdSets.get();
+    long numEntriesScannedForScanBasedDocIdSets = 0L;
+    if (scanBasedDocIdSets != null) {
+      for (BlockDocIdSet scanBasedDocIdSet : scanBasedDocIdSets) {
+        numEntriesScannedForScanBasedDocIdSets += scanBasedDocIdSet.getNumEntriesScannedInFilter();
+      }
     }
-    return _numEntriesScannedInFilter;
+    return _numEntriesScannedInFilter + numEntriesScannedForScanBasedDocIdSets;
   }
 }

